ert/experiment_server/_server.py

```diff
@@ -0,0 +1,179 @@
+import asyncio
+import logging
+import pickle
+import sys
+from contextlib import contextmanager
+from typing import TYPE_CHECKING, Any, AsyncIterator, Callable, Generator, Set
+
+from cloudevents.exceptions import DataUnmarshallerError
+from cloudevents.http import from_json
+from cloudevents.http.event import CloudEvent
+from websockets.legacy.server import WebSocketServerProtocol
+from websockets.server import serve
+from ert.serialization import evaluator_unmarshaller
+from ert_shared.ensemble_evaluator.config import EvaluatorServerConfig
+from res.enkf.enkf_main import EnKFMain
+
+from ._experiment_protocol import Experiment
+from ._registry import _Registry
+
+if sys.version_info < (3, 7):
+    from async_generator import asynccontextmanager
+else:
+    from contextlib import asynccontextmanager
+
+if TYPE_CHECKING:
+    from typing import Optional
+
+
+logger = logging.getLogger(__name__)
+event_logger = logging.getLogger("ert.event_log")
+
+
+class ExperimentServer:
+    def __init__(self, ee_config: EvaluatorServerConfig) -> None:
+        self._config = ee_config
+        self._registry = _Registry()
+        self._clients: Set[WebSocketServerProtocol] = set()
+
+        self._dispatchers_connected: "asyncio.Queue[Optional[WebSocketServerProtocol]]" = (  # noqa
+            asyncio.Queue()
+        )
+
+        self._server_done = asyncio.get_running_loop().create_future()
+        self._server_task = asyncio.create_task(self._server())
+
+    async def _handler(self, websocket: WebSocketServerProtocol, path: str) -> None:
+        elements = path.split("/")
+        if elements[1] == "client":
+            await self.handle_client(websocket, path)
+        elif elements[1] == "dispatch":
+            logger.debug("dispatcher connected")
+            await self.handle_dispatch(websocket, path)
+        else:
+            logger.info(f"Connection attempt to unknown path: {path}.")
+
+    async def stop(self) -> None:
+        logger.debug("stopping experiment server gracefully...")
+        try:
+            self._server_done.set_result(None)
+        except asyncio.InvalidStateError:
+            logger.debug("was already gracefully asked to stop.")
+            pass
+        await self._server_task
+
+    @asynccontextmanager
+    async def count_dispatcher(self) -> AsyncIterator[None]:
```

Comment:
Do we need this function in experiment server since it. I understand the reason when using it in ensemble evaluator (reconecting clients to make sure they disconnects properly), but not sure if this is the case here.
If so I would welcome a sentence or two in a comment to introduce the need for it.